src/core/util/hugepage_mgr.cpp

```diff
@@ -0,0 +1,258 @@
+/*
+ * Copyright (c) 2001-2023 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "hugepage_mgr.h"
+
+#include <sys/types.h>
+#include <dirent.h>
+#include <sys/mman.h>
+
+#include <algorithm>
+#include <cmath>
+#include <cstring>
+#include <mutex>
+#include <fstream>
+#include <sstream>
+
+#include "vlogger/vlogger.h"
+#include "util/sys_vars.h"
+
+#define MODULE_NAME "hugepage_mgr"
+
+hugepage_mgr g_hugepage_mgr;
+
+hugepage_mgr::hugepage_mgr()
+{
+    memset(&m_stats, 0, sizeof(m_stats));
+    m_default_hugepage = read_meminfo("Hugepagesize:");
+    update();
+}
+
+void hugepage_mgr::update()
+{
+    std::lock_guard<decltype(m_lock)> lock(m_lock);
+
+    m_hugepages.clear();
+    read_sysfs();
+}
+
+bool hugepage_mgr::is_hugepage_optimal(size_t hugepage, size_t size)
+{
+    return (size / hugepage) || hugepage_unused_space(hugepage, size) <= HUGEPAGE_UNUSED_OPTIMAL ||
```

Comment:
Seems like this logic is relevant only if size < hugepage. but then isn't the samllest hugepage is best?

Reply:
It's not an easy task to implement perfect optimal and acceptable criteria. We don't have enough information for this. After playing with 2 kinds of allocations simultaneously (buffer pools and buffer/tcp_seg metadata) I came with current naive approach.

Hugepages are checked in descending order, so the biggest optimal or the biggest acceptable will be used.